src/web-extensions.c

```diff
@@ -0,0 +1,861 @@
+  if (!host_path)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  xdp_dbus_web_extensions_complete_get_manifest (object, invocation, json_manifest);
+  return TRUE;
+}
+
+static void
+handle_start_in_thread (GTask *task,
+                        gpointer source_object,
+                        gpointer task_data,
+                        GCancellable *cancellable)
+{
+  Request *request = (Request *)task_data;
+  Session *session;
+  WebExtensionsSession *web_extensions_session;
+  const char *arg_name;
+  char *arg_extension_or_origin;
+  const char *app_id;
+  g_autofree char *host_path = NULL;
+  g_autofree char *description = NULL;
+  g_autofree char *manifest_filename = NULL;
+  guint response = XDG_DESKTOP_PORTAL_RESPONSE_OTHER;
+  gboolean should_close_session;
+  Permission permission;
+  gboolean allowed;
+  char *argv[] = {NULL, NULL, NULL, NULL};
+  g_autoptr(GError) error = NULL;
+
+  REQUEST_AUTOLOCK (request);
+  session = g_object_get_data (G_OBJECT (request), "session");
+  SESSION_AUTOLOCK_UNREF (g_object_ref (session));
+  g_object_set_data (G_OBJECT (request), "session", NULL);
+  web_extensions_session = (WebExtensionsSession *)session;
+
+  if (!request->exported || web_extensions_session->state != WEB_EXTENSIONS_SESSION_STATE_STARTING)
+    goto out;
+
+  arg_name = g_object_get_data (G_OBJECT (request), "name");
+  arg_extension_or_origin = g_object_get_data (G_OBJECT (request), "extension-or-origin");
+
+  host_path = find_messaging_host (web_extensions_session->mode,
+                                   arg_name, arg_extension_or_origin,
+                                   &description, &manifest_filename, NULL,
+                                   &error);
+  if (host_path == NULL)
+    {
+      g_warning ("Could not find WebExtensions backend: %s", error->message);
+      fflush(stderr);
+      fflush(stdout);
+      goto out;
+    }
+
+  app_id = xdp_app_info_get_id (request->app_info);
+  permission = get_permission_sync (app_id, PERMISSION_TABLE, arg_name);
+  if (permission == PERMISSION_ASK || permission == PERMISSION_UNSET)
+    {
+      guint access_response = 2;
+      g_autoptr(GVariant) access_results = NULL;
+      GVariantBuilder opt_builder;
+      g_autoptr(GAppInfo) info = NULL;
+      const char *display_name;
+      g_autofree gchar *app_info_id = NULL;
+      g_autofree gchar *title = NULL;
+      g_autofree gchar *subtitle = NULL;
+      g_autofree gchar *body = NULL;
+
+      info = xdp_app_info_load_app_info (request->app_info);
+      if (info)
+        {
+          g_auto(GStrv) app_id_components = g_strsplit (g_app_info_get_id (info), ".desktop", 2);
+          app_info_id = g_strdup (app_id_components[0]);
+        }
+      display_name = info ? g_app_info_get_display_name (info) : app_id;
+      title = g_strdup_printf (_("Allow %s to start WebExtension backend?"), display_name);
```

Comment:
When using a firefox instance installed like `sudo snap install --channel=latest/edge firefox_edge`, I dont see that prompt and the `NativeMessagingPortal` logging on Firefox side directly shows us cancellation `[Parent 3974: Main Thread]: D/NativeMessagingPortal native application start canceled by user in session /org/freedesktop/portal/desktop/session/1_114/firefox_org_gnome_chrome_gnome_shell_1689606647` , suggested we went through `response = XDG_DESKTOP_PORTAL_RESPONSE_CANCELLED;` below.

Reply:
The `Only the focused app is allowed to show a system access dialog` comes from the permission dialog. I'm having the same problem when running firefox from the gdb (focused terminal) or when the wmclass does not match to the `app_id`.  You can set wmclass by env variable `MOZ_APP_REMOTINGNAME`. TO check if they match I use `lg` command in Gnome Shell.

Reply:
Right, but that's **after** hacking the database. I'm unsure why there's the `no` value at first. Could it be that the remoting name is used and it does not match the aliased name, and we default to `no` ?

Reply:
The error message originates from gnome-shell, which denies the call when there is a mismatch in requesting app ID and the ID of the focused app, at https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/64d5d01360bb072ca476da85c6163f9dbef78423/js/ui/accessDialog.js#L167

Relevant discussion is also at https://github.com/flatpak/xdg-desktop-portal/issues/769.

> Manually changing (yes ...) the value to yes, I'm crashing XDG Desktop Portal

That sounds like a bug in the portal implementation. Regardless of client behavior, the portal should never crash.

Reply:
Interesting, might be related to https://github.com/canonical/firefox-snap/blob/71b01a1a1aa3dc1024cca6ec9817bee25348077d/snapcraft.yaml#L42-L43 where we force the value. One installed alias would be e.g.,
```
-rw-r--r-- 1 root root  30K déc.  10 19:32 /var/lib/snapd/desktop/applications/firefox+edge_firefox.desktop
-rw-r--r-- 1 root root 9,4K déc.   1 13:25 /var/lib/snapd/desktop/applications/firefox_firefox.desktop
```

Reply:
I just downloaded an Ubuntu 24.04.1 iso, and installed the `gnome-browser-connector` package:

```
sudo apt-get update
sudo apt-get install gnome-browser-connector
```

Then I went to the App Center and switched Firefox from stable to edge. This replaced Firefox Release with Firefox Nightly. When I looked in the Windows tab of the `lg` debugger (Alt+F2, found instructions at [debugging at gjs.guide](https://gjs.guide/extensions/development/debugging.html#looking-glass)), I see that the app is still "firefox_firefox.desktop". And when I install the [GNOME Shell integration](https://addons.mozilla.org/en-US/firefox/addon/gnome-shell-integration/) Firefox add-on, I get the permission prompt as expected.

So far, so good. I reverted Firefox back to the regular release, and tried to get two parallel snap instances. This requires an experimental pref, so I first had to run the following and reboot:

```
sudo snap set system experimental.parallel-instances=true
reboot
```
(after running enabling parallel-instances above, existing snaps were broken, and I could not even open the App Center, so this STR cannot be run of a diskless live CD, only from a persistent Ubuntu installation due to the need to reboot)

```
sudo snap install --channel=latest/edge firefox_edge
snap run firefox_edge
```

After doing this, the browser launches without the Nightly icon, and the icon is missing/generic with the name `firefox_edge`. When I open `lg` (Alt+F2, `lg`) and Evaluate `Shell.WindowTracker.get_default().focus_app.id`, the result is `window:13` (instead of `firefox_firefox.desktop`).

Based on this experiment, it looks like there is a bug with parallel snaps.

Reply:
I get various values of `app_id` depending which type of execution of the binary I use. Using my own desktop file started from Gnome  I get correct value - like org.mozilla.firefox. When starting from Konsole, I get org.kde.konsole, when using Gnome Terminal I get empty string value. The app_id is then used as a key to the permission storage record. Unfortunately setting `MOZ_APP_REMOTINGNAME` does not make a change to the `app_id`. Only changes wmclass (which on the other hand must match to the app_id when the system access dialog is shown).

Sidenote: During testing I use only `widget.use-xdg-desktop-portal.native-messaging=1` in about:config. So the flatpak nor snap is involved. That could be reason the `app_id` varies in my case.